pkg/vacuum/vacuum.go

```diff
@@ -275,25 +304,63 @@ func (e *Engine) Run(ctx context.Context) {
 		}
 		numberVacuumConnections.Set(0)
 	}()
-	// we limit ourselves to batches of 1000 chunks
-	// since we already have the advisory lock, continue to vacuum batches as needed until none left
-	for {
-		chunks, err := e.listChunks(ctx, con)
+
+	// okay, we have the advisory lock, let's do some work...
+
+	chunksToFreeze := workload{
+		name:         "compressed chunks to freeze",
+		query:        sqlListChunksToFreeze,
+		scaleWorkers: true,
+		stop:         false,
+	}
+	chunksMissingStats := workload{
+		name:         "compressed chunks missing stats",
+		query:        sqlListChunksMissingStats,
+		scaleWorkers: false,
+		stop:         false,
+	}
+
+	for i := 0; true; i++ {
+		// continue looping until both workloads were ready
+		// to stop on their last iterations
+		if chunksToFreeze.stop && chunksMissingStats.stop {
+			break
+		}
+
+		// every 3rd iteration is chunksMissingStats
+		w := &chunksToFreeze
+		if i%3 == 0 {
+			w = &chunksMissingStats
+		}
+		w.stop = false
+
+		// find some chunks to vacuum
+		chunks, err := e.listChunks(ctx, con, w.query)
 		if err != nil {
-			log.Error("msg", "failed to list chunks for vacuuming", "error", err)
-			return
+			log.Error("msg", fmt.Sprintf("failed to list %s", w.name), "error", err)
+			w.stop = true
+			continue
 		}
 		tablesNeedingVacuum.Set(float64(len(chunks)))
```

Comment:
shouldn't we add separate metrics as well to count the work for each type?

Reply:
yes. I'll do a follow-up PR

Reply:
https://github.com/timescale/promscale/pull/1812